### ISSUE TITLE
Planning: Decide how the workflow of user account creation will work for researchers

### ISSUE THREAD

**Issue by pjmonks** [MauroDataMapper/mdm-explorer#429]
# Overview

The workflow for how researchers get initial user accounts for Mauro and the SDE must be determined. This needs to happen before a researcher is able to carry out any other SDE workflow request (particularly assignment to an organisation and department), since all SDE requests are tied to a user account. This suggests that the workflow for registering to access the Mauro Data Explorer and SDE is separate to all other SDE requests.

This issue will be used to gather initial notes/requirements and to discuss how to proceed so that concrete implementation tasks can be defined. This is a "living" document and is subject to change.

# Breakdown

- [x] Identify requirements for what is needed to create/define a user account
- [x] Research how SSO identity providers will be used to manage users and authentication
- [x] Plan how the SSO identity provider will work in conjunction with Mauro/SDE to create application user accounts for Mauro Data Explorer

# Requirements

Based on the description taken from the epic #421, the following requirements for how a researcher gets a user account can be implied:

1. A researcher needs two user accounts - one for Mauro, one for the SDE.
2. A researcher will sign in to both user accounts (synchronised) via Single Sign On (SSO).
3. The SSO identify platform will be assumed to be Microsoft Azure EntraID based on what OUH use.
4. A researcher _must_ have these user accounts before they can submit any other SDE requests or make Data Specifications.
5. Before a researcher has these user accounts, they must be able to access some pages in the Mauro Data Explorer anonymously - particularly to start the registration process.
6. For the Mauro user account, the user must be assigned to the "Explorer Readers" user group to allow permissions to read the root data model.
7. SSO must allow users outside of the OUH tenant to authenticate via the identity platform (Azure EntraID), most likely using their own organisation email address.
8. An anonymous user must be able to submit a user registration request to the SDE.
9. The creation of Mauro/SDE user account records and Azure EntraID (guest) tenant accounts will be separated. In order, Mauro/SDE user accounts will be created first, followed by Azure EntraID account.
10. Mauro/SDE and EntraID user accounts will share the same email address (username) to work and match user accounts.
11. A user registration request _must_ allow no user ID to be recorded against it initially. _However_, once the user account has been created in the SDE, the request _must_ be updated to map the user to the request.
12. A user registration request cannot be marked as "completed" until the user ID of the new user has been assigned to it.

# User Registration

## When this happens

It is necessary to submit a _user registration request_ when:

- A new user does not have an account yet
- They are able to select an existing organisation and department to join.

If they cannot find an existing organisation/department, they must swap to following the _organisation registration request_ workflow - see #430.

## Sequence of Actions

Below is the intended workflow of actions to carry out as part of the user registration request:

```mermaid
flowchart        
    subgraph explorer
        anonUser[Anonymous user]
        firstTimeSignIn[First time sign-in]
        reviewRequestStatus[Review request status]
        userReadyForSde[Ready for SDE]
    end
    subgraph sde-admin
        pendingRegistration[Pending registration request]        
        assignedRegistration[Assigned registration request]
        initialUserChecks[Initial user checks]
        verifyUserOrg[Verify user/organisation]
        createSdeUser[Create Mauro/SDE user records]
        mapUserToRequest[Map user to request]
        newUserReady[New user ready]
        rejectUser[Reject and notify user]
        mapUserToOrg[Map user to organisation]
        mapUserToDept[Map user to department]

        pendingRegistration -- SDE administrator pickup --> assignedRegistration
        assignedRegistration --> initialUserChecks
        initialUserChecks -- Verified --> createSdeUser        
        createSdeUser --> mapUserToRequest
        mapUserToRequest --> newUserReady
        initialUserChecks -- Rejected --> rejectUser
        newUserReady --> verifyUserOrg
        verifyUserOrg -- Verified --> mapUserToOrg
        mapUserToOrg --> mapUserToDept
        verifyUserOrg -- Rejected --> rejectUser
    end
    subgraph azure
        addAzureAccount[Add guest account to tenant]
        authenticateUser[Authenticate user]
    end
    anonUser -- Submit registration form --> pendingRegistration
    createSdeUser --> addAzureAccount
    newUserReady -- Notify user --> firstTimeSignIn
    addAzureAccount --> newUserReady
    firstTimeSignIn -- Authenticate --> authenticateUser
    authenticateUser --> reviewRequestStatus
    reviewRequestStatus --> userReadyForSde
    mapUserToDept --> userReadyForSde
```

## Request State Workflow

Below are some changes to the existing SDE request state workflow that will need to be achieved to support the additional request types for user registration:

```mermaid
stateDiagram-v2
    draft : Draft request
    awaitUserCheck : Await user checks
    initialUserCheck : Initial user check
    createUserAccount : Create user account
    awaitApproval : Await Approval
    awaitPickup : Await pickup
    otherSdeStates : Other SDE states...

    draft --> awaitUserCheck : Submit (Researcher)

    awaitUserCheck --> initialUserCheck : Assign (SDE Administrator)
    initialUserCheck --> draft : Rejected (SDE Administrator)
    initialUserCheck --> createUserAccount : Approved (SDE Administrator)
    createUserAccount --> awaitApproval : User ready (SDE Administrator)
    createUserAccount --> initialUserCheck : Pending (SDE Administrator)
    createUserAccount --> awaitUserCheck : Pending (SDE Administrator)

    awaitApproval --> awaitPickup : Approve (Organisation Member)
    awaitApproval --> draft : Returned (Organisation Member)

    awaitPickup --> otherSdeStates : Continue... (SDE Administrator)

```

The key differences are what happens between "Draft Request" and "Await Approval" or "Await Pickup":

- For this kind of request, the status transitions straight from "Draft Request" to "Await User Checks". 
- Only SDE administrators can manage requests in the "Await User Checks" state.
- An SDE administrator must be assigned to the request, moving to the "Initial User Checks" state.
- In this state, the SDE administrator must make some initial background checks on the user to ensure they should have a user account. (These are processes to be determined by OUH).
- All being well, they move the request to the "Create User Account" state. Here, they will perform the technical actions to setup their user account. Once created, the request can finally be mapped to this new user account.
- In the case of a user registering to join an existing organisation, this means they transition to the already available "Await Approval" state. From here onwards, the current request workflow continue as before.

## Registration Form

An anonymous user will be able to fill in a registration form in the Mauro Data Explorer, providing the following details:

- **Name of user** - first and last
- **Email address** - their organisation email address
- **Confirmations** - e.g. "I have completed online training", "I accept the terms and conditions** etc
- **Additional information** - ORCID, GMC number, NMC number etc
- **Organisation** - select an existing organisation from a list
- **Department** - select an existing department from a list based on chosen Organisation
- **Position/job title** - for records

If an Organisation or Department cannot be found from the lists, a different registration form must be submitted - the _Organisation Registration Form_ (#430)

Once submitted, the request will not currently be attached to a user ID - because the user does not exist yet. When the user record is available though, it will be assigned to the request for tracking purposes.

## SDE Administrator Actions

Once the user registration request is picked up by an SDE administrator, the following actions must happen before the user can at a minimum sign-in to the Mauro Data Explorer and track their request:

1. **Create Mauro/SDE user accounts** - the SDE administrators must perform some initial checks to make sure a user account should be created e.g. the request is not spam or a bot. Once verified, some processes can happen to setup accounts. Note that these are no longer _automatic_, they must be triggered _manually_ through the SDE Admin UI.
    1. **Mauro** - create a user account with the same email address. Also assign the user to the correct user groups.
    2. **SDE** - create a user account with the same email address.
2. **Create user in Azure** - next, a user account should be added/created in the Azure OUH tenant. This is a step performed by the OUH system administrators (currently not automated).
3. **Notification** - the user must be notified (somehow) that they can sign-in to the Mauro Data Explorer.
4. **Verification** - the SDE administrators/team must verify the suggested user and make sure they are part of that organisation - "due diligence". These are not necessarily steps that can be automated in the SDE backend.

**Note:** although the user may sign-in finally and track their request, they must still be prevented from performing any actions until they are added to an Organisation/Department.

## Add to Organisation/Department

Whilst the research user is able to access the Mauro Data Explorer, the user registration request can continue to be fulfilled by the SDE administrator. Once verified and approved, the SDE administrator will be able to run the following commands:

1. Add user to the chosen organisation
2. Add user to the chosen department within the organisation

Only once these are done can the user registration request be considered "Complete". Now the researcher is able to submit further requests e.g. creating projects.

# Identity Management

> These notes are provided as an appendix (for reference)

Some research has been carried out to understand Microsoft Azure EntraID and determine how users are managed. Ultimately, this should be handled and business processes decided by OUH - these notes may be useful though.

**Note:** The below topics refer to Azure EntraID, though I am sure the concepts could apply to any suitable identity platform.

Azure works with _tenants_, which are dedicated instances of Microsoft EntraID that manage an organisation's resources, including registered apps and a directory of users. For example, OUH would have a tenant that manages all users under the Oxford University Hospital trust organisation, giving each user an `@ouh.nhs.uk` domain email account for sign in. This is assumed to be a **workforce** tenant, the default tenant configuration.

## Collaboration with guests

It must be assumed that researchers are not solely found within the OUH tenant/organisation though, there are other external organisations (let's say _Cambridge University Hospitals (CUH)_ trust, `@cuh.nhs.uk`) which will require access so user accounts and authentication must work for these too. Microsoft EntraID therefore provides many powerful capabilities for external "B2B" collaboration. This will allow OUH to invite external business partners as guest accounts within the OUH tenant, or to provide self-service capabilities to let users sign-up themselves.

Further documentation can be found here:

- [Introduction to Microsoft External Entra External ID](https://learn.microsoft.com/en-gb/entra/external-id/external-identities-overview)
- [Overview: B2B collaboration with external guests for your workforce](https://learn.microsoft.com/en-gb/entra/external-id/what-is-b2b)
- [Quickstart: Add a guest user and send an invitation](https://learn.microsoft.com/en-gb/entra/external-id/b2b-quickstart-add-guest-users-portal)
- [Identity providers for External ID in workforce tenants](https://learn.microsoft.com/en-gb/entra/external-id/identity-providers)
- [Self-service sign-up](https://learn.microsoft.com/en-gb/entra/external-id/self-service-sign-up-overview)

## Options for user accounts in Azure

Defining the business processes and workflows for managing all SDE users within Azure EntraID is outside the scope of this project, but would be worth discussing with OUH to understand how researchers could start to get user accounts created in Mauro/SDE. From the research above, some of these options seem viable:

1. Azure administrators create guest user accounts in the OUH tenant for external users, using their own work email account. This will send the external user an email invitation to accept and validate.
6. Self-service sign-up could allow external users to add themselves as guests to the tenant, removing the need for OUH administrators to manually administrate guest accounts and send invites.
7. _Identity providers_ in EntraID can allow external users to authenticate using their own external accounts to complete B2B collaboration invitations or complete sign-up. This is particularly useful for other organisations using EntraID, since this is the default identity provider - so if OUH and CUH both use EntraID, a CUH user can simply use their `@cuh.nhs.uk` account as usual without any friction. There are also other identity providers available, e.g. Google.

## App Registration

Assuming that the OUH tenant contains users (either internal or external), an EntraID App Registration is required to allow Mauro and the SDE to connect to the OUH tenant and authenticate/authorize user accounts to allow further access. 

The details of App Registrations can be summarised by this page: [Application and service principal objects in Microsoft EntraID](https://learn.microsoft.com/en-us/entra/identity-platform/app-objects-and-service-principals?tabs=browser). The basics are:

1. A tenant will have one or more _applications_ registered and integrated. This will be a globally unique instance that lives within the tenant, and will allow other client applications to send authentication requests to be verified by EntraID. These are referred to as the _application object_.
5. Each _application object_ also has a _service principal_. This defines what the app can actually do in the specific tenant, who can access the app and what resources the app can access.

In the Azure Portal or Microsoft EntraID admin centre under your EntraID tenant, these will appear as:

1. **App registrations** - the _application object_, determining Client ID and Secret for client applications to connect to for OpenID Connect to work from Mauro/SDE
4. **Enterprise applications** - the _service principal_, determining which users or groups of users could access the _app registration_.

Using this knowledge, Mauro and the SDE would have an app registration devoted to SDE researcher users, where Mauro and the SDE researcher API would authenticate to. To further secure this authentication endpoint, the service principal can be managed to a set of users who are only meant to be "researchers" - i.e. SDE administrators or other OUH users not using the SDE can be excluded.

This is still outside of the scope of the Mauro/SDE project. Nonetheless, it will prove as useful background information to understand the user registration and authentication process that must be determined.

## SDE Administrators

So far, SDE researchers have been considered. However, SDE administrators will also need a user account for SSO. The difference here is that they will only need one client account for the SDE Admin UI, they do not require a Mauro user account. 

Some consideration needs to be given to understand how SDE administrators register or are provided a user account to access the SDE Admin UI.

**Comment by pjmonks** [comment]
@jamesrwelch and @NigelPalmer - I have documented what I've understood about Microsoft EntraID and user management above. In order to plan the work for researcher user registration in the SDE, I will need to understand what OUH plan to do with user/identity management. Ultimately, users accounts must somehow be added to Azure EntraID - the question is how? Some of that falls outside of our scope, but we need to understand your (intended) processes so that we can connect up to it.

**Comment by pjmonks** [comment]
@jamesrwelch and @NigelPalmer - I have changed the requirements notes for user registration. Please could you review the workflow I've created - this I think will work to split up user account creation and tracking the request, though it does mean that we need a `Request` object that can initially have a `userId` of `null` attached.

This leads me to some questions I'll need clarifying to know that this approach will work (cc @abwilson23):

1. Do we alter the base `Request` object to say `createdBy` (user UUID) is nullable? Or do we create a new special type of anonymous user request? I need to understand the code infrastructure more.
2. Do we continue using the same request workflow used by all other requests? I need to understand how the state flow works and how actions happen within the workflow.
3. We need a way of notifying users that they can now sign-in, does the SDE have anything like that yet?
4. Will the SDE Admin UI now need additional screens/views for proper user management? The user registration request could automate the creation of the user accounts (similar to how the `CreateProjectRequest` works), but there are currently no controls for viewing/managing the full user list in the SDE.
5. I've left the question of how SDE administrators get user accounts open - this has not been covered anywhere yet.

**Comment by pjmonks** [comment]
I will close this issue as it is now superseded by #445.
